src/Hexagons.jl

```diff
@@ -19,177 +24,326 @@ immutable HexagonAxial <: Hexagon
     r::Int
 end
 
-
 immutable HexagonCubic <: Hexagon
     x::Int
     y::Int
     z::Int
 end
 
-
 immutable HexagonOffsetOddR <: Hexagon
     q::Int
     r::Int
 end
 
-
 immutable HexagonOffsetEvenR <: Hexagon
     q::Int
     r::Int
 end
 
 
-# Neighbor hexagon iterator
-# -------------------------
-
-immutable HexagonCubicNeighborIterator
-    hex::HexagonCubic
-end
-
-const cubic_hex_neighbor_offsets =
-    [ 1 -1  0
-      1  0 -1
-      0  1 -1
-     -1  1  0
-     -1  0  1
-      0 -1  1 ]
-
+# Basic constructors
+# ------------------
 
-function neigbors(hex::HexagonCubic)
-    HexagonCubicNeighborIterator(hex)
-end
-
-
-function start(it::HexagonCubicNeighborIterator)
-    return 1
-end
-
-
-function next(it::HexagonCubicNeighborIterator, state)
-    neighbor = HexagonCubic(it.hex.x + cubic_hex_neighbor_offsets[state, 1],
-                            it.hex.y + cubic_hex_neighbor_offsets[state, 2],
-                            it.hex.z + cubic_hex_neighbor_offsets[state, 3])
-    return (neighbor, state+1)
-end
-
-
-function done(it::HexagonCubicNeighborIterator, state)
-    return state > size(cubic_hex_neighbor_offsets, 1)
-end
+hexagon(x::Int, y::Int, z::Int) = HexagonCubic(x, y, z)
+hexagon(q::Int, r::Int) = HexagonAxial(q, r)
 
 
 # Convert between hexagon indexing
 # --------------------------------
 
-
 function convert(::Type{HexagonAxial}, hex::HexagonCubic)
-    return HexagonAxial(hex.x, hex.z)
+    HexagonAxial(hex.x, hex.z)
 end
 
-
 function convert(::Type{HexagonAxial}, hex::HexagonOffsetOddR)
-    return convert(HexagonAxial, convert(HexagonCubic, hex))
+    convert(HexagonAxial, convert(HexagonCubic, hex))
 end
 
-
 function convert(::Type{HexagonAxial}, hex::HexagonOffsetEvenR)
-    return convert(HexagonAxial, convert(HexagonCubic, hex))
+    convert(HexagonAxial, convert(HexagonCubic, hex))
 end
 
-
 function convert(::Type{HexagonCubic}, hex::HexagonAxial)
-    return HexagonCubic(hex.q, hex.r, -hex.q - hex.r)
+    HexagonCubic(hex.q, hex.r, -hex.q - hex.r)
 end
 
-
 function convert(::Type{HexagonCubic}, hex::HexagonOffsetOddR)
-    x = hex.q - div(hex.r - (hex.r & 1), 2)
+    x = hex.q - (hex.r >> 1)
```

Comment:
hex.r & is one iff hex.r is odd. When you shift one bit to the right, the least significant bit is dropped anyways, so this comes down to the same thing (and is faster if the compiler doesn't optimize it away)